src/vm/jitinterface.cpp

```diff
@@ -8380,37 +8380,47 @@ CorInfoIntrinsics CEEInfo::getIntrinsicID(CORINFO_METHOD_HANDLE methodHnd,
     else
     {
         MethodTable * pMT = method->GetMethodTable();
-        if (pMT->GetModule()->IsSystem() && pMT->IsByRefLike())
+        if (pMT->GetModule()->IsSystem())
         {
-            if (pMT->HasSameTypeDefAs(g_pByReferenceClass))
+            if (pMT->IsByRefLike())
             {
-                // ByReference<T> has just two methods: constructor and Value property
-                if (method->IsCtor())
-                {
-                    result = CORINFO_INTRINSIC_ByReference_Ctor;
-                }
-                else
+                if (pMT->HasSameTypeDefAs(g_pByReferenceClass))
                 {
-                    _ASSERTE(strcmp(method->GetName(), "get_Value") == 0);
-                    result = CORINFO_INTRINSIC_ByReference_Value;
+                    // ByReference<T> has just two methods: constructor and Value property
+                    if (method->IsCtor())
+                    {
+                        result = CORINFO_INTRINSIC_ByReference_Ctor;
+                    }
+                    else
+                    {
+                        _ASSERTE(strcmp(method->GetName(), "get_Value") == 0);
+                        result = CORINFO_INTRINSIC_ByReference_Value;
+                    }
+                    if (pMustExpand != nullptr)
+                    {
+                        *pMustExpand = true;
+                    }
                 }
-                if (pMustExpand != nullptr)
+                else if (pMT->HasSameTypeDefAs(MscorlibBinder::GetClass(CLASS__SPAN)))
                 {
-                    *pMustExpand = true;
+                    if (method->HasSameMethodDefAs(MscorlibBinder::GetMethod(METHOD__SPAN__GET_ITEM)))
+                    {
+                        result = CORINFO_INTRINSIC_Span_GetItem;
+                    }
                 }
-            }
-            else if (pMT->HasSameTypeDefAs(MscorlibBinder::GetClass(CLASS__SPAN)))
-            {
-                if (method->HasSameMethodDefAs(MscorlibBinder::GetMethod(METHOD__SPAN__GET_ITEM)))
+                else if (pMT->HasSameTypeDefAs(MscorlibBinder::GetClass(CLASS__READONLY_SPAN)))
                 {
-                    result = CORINFO_INTRINSIC_Span_GetItem;
+                    if (method->HasSameMethodDefAs(MscorlibBinder::GetMethod(METHOD__READONLY_SPAN__GET_ITEM)))
+                    {
+                        result = CORINFO_INTRINSIC_ReadOnlySpan_GetItem;
+                    }
                 }
             }
-            else if (pMT->HasSameTypeDefAs(MscorlibBinder::GetClass(CLASS__READONLY_SPAN)))
+            else if (pMT->HasSameTypeDefAs(MscorlibBinder::GetClass(CLASS__ENUM)))
```

Comment:
The preferred way to define non-generic intrinsincs is to make them FCall and use FCIntrinsic to get them ID assigned in `vm\ecalllist.h`. The advantage of it is that the intrinsic ID is cached and the scheme scales well with the number of intrinsics.

For `Enum.HasFlag`, it would require moving the bit of error handling from C# to C++ so that the main entrypoint is FCall.

Reply:
So basically move the checking in `HasFlag` down into `InternalHasFlag`, then remove the former and rename the latter as `HasFlag`?

Reply:
Right

Reply:
Looked into this some -- it is not obvious how to create the proper exception with formatted text for the class equivalence check. Any suggestions?